Add post Write-Up of the PyConDE & PyData Berlin 2019 conference (#21)
JDASoftwareGroup/jdatech.github.io: _posts/2020-01-26-Pycon-de-writeup.markdown

---
layout: single
title:  "Write-Up of the PyConDE & PyData Berlin 2019 conference"
date:   2020-01-26 20:00:00 +0100
tags: conference python talk
header:
  overlay_image: assets/images/background-banner-blur-conference-399158.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false
author: Sebastian Neubauer
author_profile: true
---
# Write-Up of the PyConDE & PyData Berlin 2019 conference

The CfP for the 2020 PyCon.DE & PyData Berlin conference is now open (submit your proposal 
[here](https://de.pycon.org/)). This is a great time to share our experiences from the 2019 conference and compile a 
list of our highlight talks. Take these as inspirations for your own submission.

PyCon.DE is a Python conference held in Germany where attendees can meet to learn about new and upcoming Python 
libraries, tools, software, and data science.

> In 2019 PyCon.DE & PyData Berlin took place in Berlin and had more than 1000 participants basically doubling the amount 

Reviewer: ```suggestion
In 2019 PyCon.DE & PyData Berlin took place in Berlin and had more than 1000 participants, basically doubling the amount 
```

> of attendees as compared to the 2018 conference in Karlsruhe. As almost all Python conferences this event was run and organised by 

Reviewer: ```suggestion
of attendees as compared to the 2018 conference in Karlsruhe. As almost all Python conferences, this event was run and organised by 
```

volunteers contributing free time and supported by community sponsors contributing work time of their employees. 
Sebastian Neubauer from Blue Yonder is part of the organizing team.

> Python is basically used throughout the entire Blue Yonder stack and attending this conference allows to stay on top of 

Reviewer: ```suggestion
Python is used throughout the entire Blue Yonder stack and attending this conference allows to stay on top of 
```

latest developments in the Python world. Furthermore, since Blue Yonder profits a lot from the open source community, 
the company was a Gold sponsor at this year's conference and thus present with a booth at the venue. In total, eleven 
Blue Yonder associates from the German offices in Karlsruhe and Hamburg traveled to the conference, talked to colleagues 

> and potential candidates at the booth, and learned new things attending the talks. Furthermore, three Blue Yonder 

Reviewer 2: Second time in a row using _Furthermore_

Reviewer: ```suggestion
and potential candidates at the booth, and learned new things attending the talks. Additionally, three Blue Yonder 
```

associates gave a talk themselves:

 * Florian Jetter: **"Kartothek – Table management for cloud object stores powered by Apache Arrow and Dask"**  
    Link: [https://www.youtube.com/watch?v=8wAH52rlADI](https://www.youtube.com/watch?v=8wAH52rlADI)

 * Sebastian Neubauer: **"6 Years of Docker: The Good, the Bad and Python Packaging"**  
    Link: [https://www.youtube.com/watch?v=kgqdOftkZ-E](https://www.youtube.com/watch?v=kgqdOftkZ-E)

 * Stefan Maier: **"Embrace uncertainty! Why to go beyond point estimators for valuable ML applications"**  
    Link: [https://www.youtube.com/watch?v=Ght8AVFxBwk](https://www.youtube.com/watch?v=Ght8AVFxBwk)
 
 
## Highlights

> There were so many good talks, so here is a very incomplete compilation of our highlights in random order. Thanks to 

Reviewer: np
```suggestion
There were many good talks, so here is a very incomplete compilation of our highlights in random order. Thanks to 
```

Stefan Maier, Sven Fritz, Vasu Sharma, Phillip Sonntag, Jakob Herpel, Aniruddh Goteti and Lucas Rademaker for their 
contribution to this list.

### Title: TBC

Link: [https://www.youtube.com/watch?v=zZXSGzlVxvU](https://www.youtube.com/watch?v=zZXSGzlVxvU)

Speaker: James Powell

> Summary: James is a "rockstar" among the python speakers, that much that he is not in the need for handing in an abstract or even a title and still being accepted. He is well known for his fast-paced presentations only using vim. This time he talked about metaprogramming in python, but there is no way around to watch it to get a clue what it is about.

Reviewer: ```suggestion
Summary: James is a "rockstar" among the python speakers, that much that he is not in the need for handing in an abstract or even a title and still being accepted. He is well known for his fast-paced presentations only using vim. This time he talked about metaprogramming in python, but there is no way around watching the talk to get a clue on what it is about.
```

Reviewer: forgot these suggestions?

Author: yap

### Title: How MicroPython went into space

Link: [https://www.youtube.com/watch?v=Vh_5Lz1mLq8](https://www.youtube.com/watch?v=Vh_5Lz1mLq8)

Speaker: Christine Spindler

> Summary: It is just mind-blowing in how many fields python drains into. Not only there is MicroPython which brings python to embedded devices, but this project now is under certification for the use in space!! So, not long until python will steer and control satellites and the space stations. Awesome!

Reviewer: ```suggestion
Summary: It is just mind-blowing in how many fields python drains into. Not only there is MicroPython which brings python to embedded devices, but this project now is under certification for the use in space!! So, it won't be long until python will steer and control satellites and the space stations. Awesome!
```

### Title: Automating feature engineering for supervised learning? Methods, open-source tools and prospects.

Link: [https://www.youtube.com/watch?v=XK7mKGWSue8](https://www.youtube.com/watch?v=XK7mKGWSue8)

Speaker: Thorben Jensen

> Summary: The talk basically reported a case study comparing three open source libraries ``tsfresh``, ``featuretools`` and ``TPOT``. At the end of the day, ``TPOT`` seemed superior in most categories, including the degree of automation. However, the case study was only carried out on a single data set. When asked about the potential biases of this dataset, the speaker said that ``tsfresh`` does provide the better support for time series problems, and that this may be more critical for other datasets.

Reviewer: another "basically"

Link: [https://www.youtube.com/watch?v=R-ts_mxGKOg](https://www.youtube.com/watch?v=R-ts_mxGKOg)

Speaker: Johannes Knopp

Summary: Johannes shows the historic evolution of their product category classification at "solute GmbH", a Karlsruhe based competitor prices service provider. I love how he openly talks about failed attempts, historic quirks and problems they faced over the years.

### Title: Dr. Schmood's Notebook of Python Calisthenics and Orthodontia

Link: [https://www.youtube.com/watch?v=BAsL_cbkQDc](https://www.youtube.com/watch?v=BAsL_cbkQDc)

Speaker: David Schmudde

Summary: I have to admit, the title would not have pulled me into this talk as I have no clue what the words are about. As I was session chair, I was lucky enough to attend it. In fact, it was about the state handling in python and why some complain about "messed up variable states" in IPython notebooks. He also suggests to embrace immutability to solve the situation and I love immutability!

### Title: Python 2020+

Link: [https://www.youtube.com/watch?v=fOdCxum-qLA](https://www.youtube.com/watch?v=fOdCxum-qLA)

Speaker: Łukasz Langa

> Summary: In this keynote, Łukasz gives a unique look behind the scenes of how Python, especially cpython, is developed and he gives some predictions where the future will, or should go. But you should definitively watch the talk because of the many anecdotes shared by him. BTW: In the hallway track, Łukasz, the creator of "black" assured me he also doesn't like some of the formatting decisions, but it is "black". Even he is not able to argue with it, it's "black"!

Reviewer: ```suggestion
Summary: In this keynote, Łukasz gives a unique look behind the scenes of how Python, especially CPython, is developed and he gives some predictions where the future will, or should go. But you should definitively watch the talk because of the many anecdotes shared by him. BTW: In the hallway track, Łukasz, the creator of "black" assured me he also doesn't like some of the formatting decisions, but it is "black". Even he is not able to argue with it, it's "black"!
```

### Title: Rethinking Open Source in the Era of Cloud & Machine Learning

Link: [https://www.youtube.com/watch?v=QMJIh-voWng](https://www.youtube.com/watch?v=QMJIh-voWng)

Speaker: Peter Wang

> Summary: Having the CEO of a big software company give the keynote at a community driven conference is at least unexpected. But Peter Wang has definitively proven that it was a very good choice. In his talk "Rethinking Open Source in the Era of Cloud & Machine Learning" he is deep-diving into how to sustainably run an open source project, commercial or uncommercial. A must see for everyone who is interested in the hidden forces behind the tectonic shifts of the  IT landscape in the recent years.

Reviewer: ```suggestion
Summary: Having the CEO of a big software company give the keynote at a community driven conference is at least unexpected. But Peter Wang has definitively proven that it was a very good choice. In his talk "Rethinking Open Source in the Era of Cloud & Machine Learning" he dives deeply into how to sustainably run an open source project, commercial or noncommercial. A must see for everyone who is interested in the hidden forces behind the tectonic shifts of the  IT landscape in the recent years.
```

### Title: Are you sure about that?! Uncertainty Quantification in AI

Link: [https://www.youtube.com/watch?v=LCDIqL-8bHs](https://www.youtube.com/watch?v=LCDIqL-8bHs)

Speaker: Florian Wilhelm

> Summary: The talk introduced the concepts of aleatoric and epistemic uncertainty. It compared various methods for uncertainty estimates according to several categories, such as performance, implementation effort etc. A simple, one variable toy dataset was used to evaluate these methods in practice. Some methods apparently showed a poor performance such as Monte-Carlo dropouts. I personally would have like to learn on why some methods performed better or worse on the dataset or not and how this generalizes to real-world datasets. However, based on later conversations with the speaker, this seems a tough problem for some of the methods used. What I definitely learned was how to give an easy explanation on the difference between aleatoric and epistemic uncertainty, and on quantile regression to a broad audience. And it was the first time I had been given such a systematic overview on uncertainty quantification.

Reviewer: ```suggestion
Summary: The talk introduced the concepts of aleatoric and epistemic uncertainty. It compared various methods for uncertainty estimates according to several categories, such as performance, implementation effort etc. A simple, one variable toy dataset was used to evaluate these methods in practice. Some methods apparently showed a poor performance such as Monte-Carlo dropouts. I personally would have like to learn on why some methods performed better or worse on the dataset or not and how this generalizes to real-world datasets. However, based on later conversations with the speaker, this seems to be a tough problem for some of the methods used. What I definitely learned was how to give an easy explanation on the difference between aleatoric and epistemic uncertainty, and on quantile regression to a broad audience. And it was the first time I had been given such a systematic overview on uncertainty quantification.
```

### Title: Time series modelling with probabilistic programming

Link: [https://www.youtube.com/watch?v=tLao3IthrfI](https://www.youtube.com/watch?v=tLao3IthrfI)

Speaker: Sean Matthews, Jannes Quer

Summary: The talk presented an extrapolation problem in demand forecasting (the aggregated demand for drugs). It was remarkably different from many other data science talks in several respects.

> 1.  It did rather deal with seemingly old-school methods on a small dataset.

Reviewer: rather?

3.  The speaker was extremely explicit on the methods chosen and about the implementation, although the problem was not an academic one, but occurred in an industry context. (He showed parts of his `pystan` code explicitly.)
4.  The modelling was more of a one-off undertaking and not conceived for contributing to a productive model pipeline that is automatically retrained regularly. When I asked the speaker if he would recommend fitting the same model again after one year, the answer was a clear "No, since I don't know the future, I can't tell whether the model would still perform well then."

Although all this seems pretty much old school, I believe that we ML youngsters can learn a lot from the first generation data scientists: Asking whether we understand why a certain model performs best, and considering also well established methods and not only the ones at their current hype.

### Title: Automated Feature Engineering and Selection in Python

Link: [https://www.youtube.com/watch?v=4-4pKPv9lJ4](https://www.youtube.com/watch?v=4-4pKPv9lJ4)

Speaker: Franziska Horn

> Summary: The speaker gave an overview on the typical feature engineering workflow in machine learning. She illustrated how this can be automated useing the [autofeat](https://github.com/cod3licious/autofeat) library. My impression was that this is a viable workflow when you start off with a new dataset, to get a good first iteration and to gain insight into the dataset. It however it seems not to be a tool that can automate away feature engineering completely when you are aiming for best-in-class predictions with a high degree of reliability and retraceability. Nevertheless it could a huge time-saver on the way to that goal.

Reviewer: ```suggestion
Summary: The speaker gave an overview on the typical feature engineering workflow in machine learning. She illustrated how this can be automated useing the [autofeat](https://github.com/cod3licious/autofeat) library. My impression was that this is a viable workflow when you start off with a new dataset, to get a good first iteration and to gain insight into the dataset. It, however, does not seem to be a tool that can automate away feature engineering completely when you are aiming for best-in-class predictions with a high degree of reliability and retraceability. Nevertheless it could a huge time-saver on the way to that goal.
```

### Title: Law, ethics and machine learning - a curious ménage à trois

Link: [https://www.youtube.com/watch?v=FK-FmuEi5BU](https://www.youtube.com/watch?v=FK-FmuEi5BU)

Speaker: Dr. Benjamin Werthmann

Summary: First of all it really stood out of the mass of talks because Benjamin is not a Python enthusiast or even developer, but a lawyer. Benjamin gave an useful overview on GDPR (General Data Protection Regulation) and it's implications. What really made his talk stay in my mind was his call to action in the end. His message: the open source community must actively participate in topics of law and ethics in machine learning in order to get satisfying results, otherwise lawyers will do it.

### Title: Refactoring in Python: Design Patterns and Approaches

Link: [https://www.youtube.com/watch?v=ZzKaFJxiDzA](https://www.youtube.com/watch?v=ZzKaFJxiDzA)

Speaker: Tin Marković

> Summary: Tim currently works at [Kiwi.com](http://Kiwi.com) and focussed his talk on reducing "code debt" by refactoring the code base regularly. The key elements he touched upon were easy wins, like automating some checks by using `black`, `mypy`, `coala`, etc., patterns that hint to "smelly code", and possible reasons ranging from "historical reasons" to "high priority urgent hacky requests". This is often not easily apparent, however, tools like "SonarQube" can help in identifying sections of code that might need refactoring. Another element that he talked about are over use of decorators, which may seem like a good idea but can lead to non-obvious functionality, and recommended that they be only limited, and shouldn't at least alter the function signature and calls. Another suggestion that I liked was around code reviews, in which it may be good idea to structure the reviews at overall scope, followed by system scope, and finally code scope. This can help save coding time in case the architecture or overall scope needs to be changed. Overall, the focus was to always keep an eye out for code debt and try to improve little by little.

Reviewer: ```suggestion
Summary: Tim currently works at [Kiwi.com](http://Kiwi.com) and focused his talk on reducing "code debt" by refactoring the code base regularly. The key elements he touched upon were easy wins, like automating some checks by using `black`, `mypy`, `coala`, etc., patterns that hint to "smelly code", and possible reasons ranging from "historical reasons" to "high priority urgent hacky requests". This is often not easily apparent, however, tools like "SonarQube" can help in identifying sections of code that might need refactoring. Another element that he talked about are over use of decorators, which may seem like a good idea but can lead to non-obvious functionality, and recommended that they be only limited, and shouldn't at least alter the function signature and calls. Another suggestion that I liked was around code reviews, in which it may be good idea to structure the reviews at overall scope, followed by system scope, and finally code scope. This can help save coding time in case the architecture or overall scope needs to be changed. Overall, the focus was to always keep an eye out for code debt and try to improve little by little.
```

and potential candidates at the booth, and learned new things attending the talks. Additionally, three Blue Yonder 
associates gave a talk themselves:

 * Florian Jetter: **"Kartothek – Table management for cloud object stores powered by Apache Arrow and Dask"**  
    Link: [https://www.youtube.com/watch?v=8wAH52rlADI](https://www.youtube.com/watch?v=8wAH52rlADI)

 * Sebastian Neubauer: **"6 Years of Docker: The Good, the Bad and Python Packaging"**  
    Link: [https://www.youtube.com/watch?v=kgqdOftkZ-E](https://www.youtube.com/watch?v=kgqdOftkZ-E)

 * Stefan Maier: **"Embrace uncertainty! Why to go beyond point estimators for valuable ML applications"**  
    Link: [https://www.youtube.com/watch?v=Ght8AVFxBwk](https://www.youtube.com/watch?v=Ght8AVFxBwk)
 
 
## Highlights

There were many good talks, so here is a very incomplete compilation of our highlights in random order. Thanks to 
Stefan Maier, Sven Fritz, Vasu Sharma, Phillip Sonntag, Jakob Herpel, Aniruddh Goteti and Lucas Rademaker for their 
contribution to this list.

### Title: TBC

Link: [https://www.youtube.com/watch?v=zZXSGzlVxvU](https://www.youtube.com/watch?v=zZXSGzlVxvU)

Speaker: James Powell

> Summary: James is a "rockstar" among the python speakers, that much that he is not in the need for handing in an abstract or even a title and still being accepted. He is well known for his fast-paced presentations only using vim. This time he talked about metaprogramming in python, but there is no way around watching the talk to get a clue on what it is about.

Reviewer 3: s/python/Python/

### Title: How MicroPython went into space

Link: [https://www.youtube.com/watch?v=Vh_5Lz1mLq8](https://www.youtube.com/watch?v=Vh_5Lz1mLq8)

Speaker: Christine Spindler

Summary: It is just mind-blowing in how many fields python drains into. Not only there is MicroPython which brings python to embedded devices, but this project now is under certification for the use in space!! So, it won't be long until python will steer and control satellites and the space stations. Awesome!

### Title: Automating feature engineering for supervised learning? Methods, open-source tools and prospects.

Link: [https://www.youtube.com/watch?v=XK7mKGWSue8](https://www.youtube.com/watch?v=XK7mKGWSue8)

Speaker: Thorben Jensen

> Summary: The talk reported a case study comparing three open source libraries ``tsfresh``, ``featuretools`` and ``TPOT``. At the end of the day, ``TPOT`` seemed superior in most categories, including the degree of automation. However, the case study was only carried out on a single data set. When asked about the potential biases of this dataset, the speaker said that ``tsfresh`` does provide the better support for time series problems, and that this may be more critical for other datasets.

Reviewer 3: Both data set and dataset are used here. Should be consistent.
